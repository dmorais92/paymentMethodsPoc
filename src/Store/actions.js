export const ACTION_TYPES = {
  PAYMENT_METHODS: {
    GET_PM_FROM_API: 'GET_PM_FROM_API',
    GET_PM_FAILED: 'GET_PM_FAILED',
    GET_PM_SUCCESS: 'GET_PM_SUCCESS',
    UPDATE_PM: 'UPDATE_PM',
    DELETE_PM: 'DELETE_PM',
  },
};

export const getPaymentMethods = () => ({
  type: ACTION_TYPES.PAYMENT_METHODS.GET_PM_FROM_API,
});

export const getPaymentMethodsFailed = () => ({
  type: ACTION_TYPES.PAYMENT_METHODS.GET_PM_FAILED,
  payload: {
    error: 'Failed getting payment methods from API',
  },
});

export const getPaymentMethodsSuccess = (paymentMethods) => ({
  type: ACTION_TYPES.PAYMENT_METHODS.GET_PM_SUCCESS,
  payload: {
    paymentMethods,
    error: false,
  },
});

export const updatePaymentMethod = (value, prop, id) => ({
  type: ACTION_TYPES.PAYMENT_METHODS.UPDATE_PM,
  payload: {
    newValue: value,
    viewProp: prop,
    updatedPaymentMethodId: id,
  },
});

export const deletePaymentMethod = (id) => ({
  type: ACTION_TYPES.PAYMENT_METHODS.DELETE_PM,
  payload: {
    deletedPaymentMethodId: id,
  },
});

export const paymentMethodsActions = {
  getPaymentMethods,
  getPaymentMethodsSuccess,
  getPaymentMethodsFailed,
  updatePaymentMethod,
  deletePaymentMethod,
};
