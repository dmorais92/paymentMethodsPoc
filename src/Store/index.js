import { applyMiddleware, createStore } from "redux";
import logger from "redux-logger";
import thunk from "redux-thunk";
import RootReducer from "./rootReducer";

const store = createStore(
  RootReducer,
  {
    paymentMethods: {
      paymentMethodsData: []
    }
  },
  applyMiddleware(logger, thunk)
);

export default store;
