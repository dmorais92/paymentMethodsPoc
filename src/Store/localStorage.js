export const loadState = () => {
  try {
    const state = localStorage.getItem('state');
    if (state === null) {
      return undefined;
    }
    return JSON.parse(state);
  } catch (err) {
    return undefined;
  }
};
export const saveState = (state) => {
  try {
    const stringifiedState = JSON.stringify(state);
    localStorage.setItem('state', stringifiedState);
  } catch {
    return undefined;
  }
  return null;
};
