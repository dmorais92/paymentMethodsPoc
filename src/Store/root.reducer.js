import { combineReducers } from 'redux';
import payments from './paymentMethods.reducer';

const Root = combineReducers({
	payments
});

export default Root;
