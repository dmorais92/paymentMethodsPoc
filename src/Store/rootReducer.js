import { combineReducers } from "redux";
import payments from "./paymentMethodsReducer.js";

const Root = combineReducers({
  payments
});
export default Root;
