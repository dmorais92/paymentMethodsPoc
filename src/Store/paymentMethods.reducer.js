import { ACTION_TYPES } from './actions';
import getPropsFromRawData from '../Utils/getPropsFromRawData';

const MAP_PATH_TO_PROP = {
	type: 'type',
	amount: 'attributes.amount',
	benificiary: 'attributes.beneficiary_party',
	charges: 'attributes.charges_information',
	debtor: 'attributes.debtor_party',
	purpose: 'attributes.payment_purpose',
	paymentType: 'attributes.payment_type',
	currency: 'attributes.currency',
	date: 'attributes.processing_date',
	reference: 'attributes.reference'
};

export default function users(
	state = {
		paymentMethodsData: [],
		isFetchingPayments: false,
		error: ''
	},
	action
) {
	switch (action.type) {
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_FROM_API:
		return {
			...state,
			isFetchingPayments: true
		};
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_SUCCESS:
		return {
			...state,
			isFetchingPayments: false,
			paymentMethodsData: action.payload.paymentMethods.map(pm =>
				getPropsFromRawData(pm, MAP_PATH_TO_PROP)
			)
		};
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_FAILED:
		return {
			...state,
			isFetchingPayments: false,
			error: action.payload.error,
			paymentMethodsData: []
		};
	default:
		return state;
	}
}
