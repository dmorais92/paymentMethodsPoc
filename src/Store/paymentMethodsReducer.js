import { ACTION_TYPES } from './actions';

export default function users(
	state = {
		paymentMethodsData: [],
		isFetchingPayments: false,
		error: ''
	},
	action
) {
	switch (action.type) {
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_FROM_API:
		return {
			...state,
			isFetchingUsers: true
		};
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_SUCCESS:
		return {
			...state,
			isFetchingUsers: false,
			userList: action.users
		};
	case ACTION_TYPES.PAYMENT_METHODS.GET_PM_FAILED:
		return {
			...state,
			isFetchingUsers: false,
			error: action.error
		};
		/*
      case ACTION_TYPES.PAYMENT_METHODS.UPDATE_PM:
        return {
          ...state,
        };
      case ACTION_TYPES.PAYMENT_METHODS.DELETE_PM:
        return {
          ...state,
        };
        */
	default:
		return state;
	}
}
