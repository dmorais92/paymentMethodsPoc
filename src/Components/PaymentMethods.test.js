import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PaymentMethodsContainer, { Payments as PaymentMethodsComponent } from './PaymentMethods';
import MOCK_DATA from '../dev.json';
import PaymentsReducer from '../Store/payments.reducer';
import { paymentMethodsActions } from '../Store/actions';

const configureMockStore = configureStore();
const { getPaymentMethodsSuccess } = paymentMethodsActions;

const INITIAL_STATE = {
  payments: {
    paymentMethodsData: [],
    isFetchingPayments: false,
    error: '',
  },
};
const UPDATED_STATE = PaymentsReducer(
  INITIAL_STATE.payments,
  getPaymentMethodsSuccess(MOCK_DATA.data),
);

const store = configureMockStore(INITIAL_STATE);
const updatedStore = configureMockStore({
  payments: UPDATED_STATE,
});

describe('PaymentMethods container', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<PaymentMethodsContainer store={store} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('matches snapshot', () => {
    const component = renderer.create(<PaymentMethodsContainer store={store} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should have the same amount of props and store payment methods', () => {
    const component = renderer.create(<PaymentMethodsContainer store={updatedStore} />);
    const componentInstance = component.root;
    expect(componentInstance.props.store.getState().payments.paymentMethodsData.length).toBe(14);
  });
});

describe('Payments component', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<PaymentMethodsComponent />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has same number of children as paymentMethods', () => {
    const component = renderer.create(
      <PaymentMethodsComponent paymentMethods={UPDATED_STATE.paymentMethodsData} />,
    );
    const testInstance = component.root;
    expect(testInstance.findAll((el) => el.type.target === 'li').length).toBe(14);
  });
});
