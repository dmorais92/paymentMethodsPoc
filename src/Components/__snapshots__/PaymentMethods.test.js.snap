// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar component matches snapshot 1`] = `
<div
  className="sc-htpNat csDlPS"
/>
`;

exports[`Button component matches snapshot 1`] = `
<button
  className="sc-dnqmqq bBTZLR"
/>
`;

exports[`Layout component matches snapshot 1`] = `
<div
  className="sc-bwzfXH bHMTeK"
/>
`;

exports[`List component matches snapshot 1`] = `
<ul
  className="sc-ifAKCX kxovTQ"
/>
`;

exports[`ListItem component matches snapshot 1`] = `
<li
  className="sc-EHOje gXHTYI"
/>
`;

exports[`Page component matches snapshot 1`] = `
<div
  className="sc-bdVaJa fPdLJN"
/>
`;
